File: exercise/php/day15/tests/BusinessTest.php
<?php

use SantaChristmasList\Business;
use SantaChristmasList\Dependencies\Factory;
use SantaChristmasList\Dependencies\Inventory;
use SantaChristmasList\Dependencies\WishList;
use SantaChristmasList\Models\Child;
use SantaChristmasList\Models\Gift;
use SantaChristmasList\Models\ManufacturedGift;

it('should load a gift into the sleigh', function () {
    $factory = new Factory();
    $inventory = new Inventory();
    $wishList = new WishList();

    $john = new Child('John');
    $toy = new Gift('Toy');
    $manufacturedGift = new ManufacturedGift('123');

    $wishList->set($john, $toy);
    $factory->set($toy, $manufacturedGift);
    $inventory->set('123', $toy);

    $business = new Business($factory, $inventory, $wishList);
    $sleigh = $business->loadGiftsInSleigh($john);

    expect($sleigh->offsetGet($john->getName()))->toBe('Gift: Toy has been loaded!');
});

it('should not load a gift if the child is not on the wish list', function () {
    $factory = new Factory();
    $inventory = new Inventory();
    $wishList = new WishList();

    $john = new Child('John');

    $business = new Business($factory, $inventory, $wishList);
    $sleigh = $business->loadGiftsInSleigh($john);

    expect($sleigh->offsetExists($john->getName()))->toBeFalse();
});

it('should not load a gift if the toy was not manufactured', function () {
    $factory = new Factory();
    $inventory = new Inventory();
    $wishList = new WishList();

    $john = new Child('John');
    $toy = new Gift('Toy');

    $wishList->set($john, $toy);

    $business = new Business($factory, $inventory, $wishList);
    $sleigh = $business->loadGiftsInSleigh($john);

    expect($sleigh->offsetExists($john->getName()))->toBeFalse();
});

it('should not load a gift if the toy was misplaced', function () {
    $factory = new Factory();
    $inventory = new Inventory();
    $wishList = new WishList();

    $john = new Child('John');
    $toy = new Gift('Toy');
    $manufacturedGift = new ManufacturedGift('123');

    $wishList->set($john, $toy);
    $factory->set($toy, $manufacturedGift);

    $business = new Business($factory, $inventory, $wishList);
    $sleigh = $business->loadGiftsInSleigh($john);

    expect($sleigh->offsetExists($john->getName()))->toBeFalse();
});

File: exercise/C#/day15/SantaChristmasList.Operations.Test/BusinessTest.cs
namespace SantaChristmasList.Operations.Test;

public class BusinessTest
{
    private const string BarCode = "123";

    private readonly Factory _factory = new();
    private readonly Inventory _inventory = new();
    private readonly WishList _wishList = new();

    private readonly Child _john = new("John");
    private readonly Gift _toy = new("Toy");
    private readonly ManufacturedGift _manufacturedGift = new(BarCode);

    [Fact]
    public void Gift_ShouldBeLoadedIntoSleigh()
    {
        _wishList.Add(_john, _toy);
        _factory.Add(_toy, _manufacturedGift);
        _inventory.Add(BarCode, _toy);

        var sut = new Business(_factory, _inventory, _wishList);
        var sleigh = sut.LoadGiftsInSleigh(_john);

        sleigh[_john].Should().Be("Gift: Toy has been loaded!");
    }

    [Fact]
    public void Gift_ShouldNotBeLoaded_GivenChildIsNotOnWishList()
    {
        var sut = new Business(_factory, _inventory, _wishList);
        var sleigh = sut.LoadGiftsInSleigh(_john);

        sleigh.ContainsKey(_john).Should().BeFalse();
    }

    [Fact]
    public void Gift_ShouldNotBeLoaded_GivenToyWasNotManufactured()
    {
        _wishList.Add(_john, _toy);
        var sut = new Business(_factory, _inventory, _wishList);
        var sleigh = sut.LoadGiftsInSleigh(_john);

        sleigh.ContainsKey(_john).Should().BeFalse();
    }

    [Fact]
    public void Gift_ShouldNotBeLoaded_GivenToyWasMisplaced()
    {
        _wishList.Add(_john, _toy);
        _factory.Add(_toy, _manufacturedGift);
        var sut = new Business(_factory, _inventory, _wishList);
        var sleigh = sut.LoadGiftsInSleigh(_john);

        sleigh.ContainsKey(_john).Should().BeFalse();
    }
}